### DISCUSSION TITLE
Tooling for migrating from version 3 to 4

### DISCUSSION THREAD

**Discussion by kasperbolarsen** [microsoft-search/pnp-modern-search#3247] [General]
I wonder how you managed to migrate/upgrade from v3 to v4. We have a fair number of customers that have used v3 extensively ( X thousand site Collections) and have now been asked to upgrade their sites and pages to v4.

So far I have identified two prereqs: a script that can identify which site collections that has the v3 app installed, and a script that lists each and every page that contains one of the v3 web parts. 

More to come. and feel free to add your own findings and ideas

**Comment by wobba** [comment]
@sympmarc did you work on something for this?

**Reply by sympmarc** [reply]
I did with your help, @wobba. Here's my blog post: https://sympmarc.com/2021/04/06/upgrading-the-pnp-modern-search-web-parts-from-v3-to-v4-where-are-they/

**Reply by kasperbolarsen** [reply]
Did you find the WebPartIds to be the same in all 3.* versions? Yes, I am paranoid.

**Reply by sympmarc** [reply]
I did. I used this script in several tenants at the time, and I found 100% of the Web Parts.

**Reply by wobba** [reply]
Maybe we could copy your code @sympmarc  (with credit and link) and create a doc page?

**Reply by sympmarc** [reply]
Absolutely. One for all and all for one.

**Reply by kasperbolarsen** [reply]
Perfect, that will take care of the generic part.  And leave me more time for the day time job part; scripting the upgrade from v3 to v4 on some 50-60K site collections on a handful of tenants 😊

**Reply by sympmarc** [reply]
Wow, that's a lot.

**Reply by paul-il779** [reply]
> Perfect, that will take care of the generic part. And leave me more time for the day time job part; scripting the upgrade from v3 to v4 on some 50-60K site collections on a handful of tenants 😊

What process were you following? How did you deal with transferring configuration (if at all)? How did you handle Handlebar differences between V3 and V4? Thank you.

**Reply by kasperbolarsen** [reply]
I had to scan the tenant, using the script below, and group the results. In many cases the config was identical on hundreds of pages, and thus I just had to recreate the config in v4 on one page, and copy it to the remaining pages using PowerShell. The unique ones I had to recreate in v4 by hand.

**Reply by sympmarc** [reply]
Good opportunity to centralize the template files, if that's possible given your usage.

**Comment by kasperbolarsen** [comment]
I rewrote Marcs script to a tenant wide scope and the PR for PnP Script Samples is https://github.com/pnp/script-samples/pull/583

Final link: https://pnp.github.io/script-samples/pnp-modern-searchv3-scanner/README.html?tabs=pnpps
